File: rust/sdk/src/types/transaction.rs
use crate::account::AccountId;
use crate::error::M10Error;
use crate::transaction_ext::TransactionExt;
use crate::{AccountUpdate, Action, Transfer, TxnFilter, DEFAULT_TXN_LIMIT};
use m10_protos::sdk;

#[derive(Debug, serde::Serialize)]
pub enum Transaction {
    Transfer(Transfer),
    InitiateTransfer(Transfer),
    CommitTransfer(Transfer),
    AccountUpdate(AccountUpdate),
    // TODO @sadroeck - fixme
    DocumentOperations,
    Action(Action),
}

impl TryFrom<sdk::FinalizedTransaction> for Transaction {
    type Error = M10Error;

    fn try_from(txn: sdk::FinalizedTransaction) -> Result<Self, Self::Error> {
        use sdk::transaction_data::Data;
        let tx = match txn.data().ok_or(M10Error::InvalidTransaction)? {
            Data::Transfer(_) | Data::InitiateTransfer(_) | Data::CommitTransfer(_) => {
                Transaction::Transfer(Transfer::try_from(txn)?)
            }
            Data::CreateLedgerAccount(_)
            | Data::SetFreezeState(_)
            | Data::SetInstrument(_)
            | Data::SetBalanceLimit(_) => Transaction::AccountUpdate(AccountUpdate::try_from(txn)?),
            Data::InvokeAction(_) => Transaction::Action(Action::try_from(txn)?),
            // TODO @sadroeck - fixme
            Data::DocumentOperations(_) => Transaction::DocumentOperations,
        };
        Ok(tx)
    }
}

pub struct ContextFilter(pub Vec<u8>);

impl TxnFilter<ContextFilter> {
    pub fn context_id(id: Vec<u8>) -> Self {
        Self {
            filter: ContextFilter(id),
            min: 0,
            max: u64::MAX,
            limit: DEFAULT_TXN_LIMIT,
        }
    }
}

impl From<TxnFilter<ContextFilter>> for sdk::ListTransactionsRequest {
    fn from(filter: TxnFilter<ContextFilter>) -> Self {
        sdk::ListTransactionsRequest {
            context_id: filter.filter.0,
            min_tx_id: filter.min,
            max_tx_id: filter.max,
            limit: filter.limit,
        }
    }
}

pub struct GroupingFilter {
    account_id: AccountId,
}

impl GroupingFilter {
    pub fn account(account_id: AccountId) -> Self {
        Self { account_id }
    }
}

impl From<TxnFilter<GroupingFilter>> for sdk::GroupTransactionsRequest {
    fn from(filter: TxnFilter<GroupingFilter>) -> Self {
        sdk::GroupTransactionsRequest {
            account_id: filter.filter.account_id.to_vec(),
            min_tx_id: filter.min,
            max_tx_id: filter.max,
            limit_groups: filter.limit,
        }
    }
}

impl TxnFilter<GroupingFilter> {
    pub fn account(account_id: AccountId) -> Self {
        Self {
            filter: GroupingFilter { account_id },
            min: 0,
            max: u64::MAX,
            limit: DEFAULT_TXN_LIMIT,
        }
    }
}

File: node/m10-sdk/src/types/transaction.ts
/* eslint-disable @typescript-eslint/no-shadow */
import { m10 } from "../../protobufs";
import type { ITxnFilter } from "../builders/txn_filter";
import { TxnFilter } from "../builders/txn_filter";
import { M10Error } from "../error";
import type { AccountId, TxId } from "../ids";
import { unwrap } from "../utils";

import { AccountUpdate } from "./account";
import { Action } from "./action";
import { Transfer } from "./transfer";


// type TransactionTypeString =  "transfer" | "createLedgerAccount" | "setFreezeState" | "documentOperations" |
//     "invokeAction" | "initiateTransfer" | "commitTransfer" | "setInstrument" | "setBalanceLimit";

export enum TransactionType {
    Transfer,
    InitiateTransfer,
    CommitTransfer,
    AccountUpdate,
    // TODO: Not yet implemented
    DocumentOperations,
    Action,
}

type TransactionValue = Transfer | AccountUpdate | Action;

export interface ITransaction {
    type: TransactionType;
    value: Option<TransactionValue>;
}

export class Transaction implements ITransaction {

    public type: TransactionType;
    public value: Option<TransactionValue>;

    public constructor(properties: ITransaction) {
        this.type = properties.type;
        this.value = properties.value;
    }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     */
    public static tryFrom(txn: m10.sdk.IFinalizedTransaction): Transaction {
        const txnData = new m10.sdk.transaction.TransactionData(
            unwrap(txn.request?.data, M10Error.InvalidTransaction()),
        );
        switch (unwrap(txnData.data, M10Error.InvalidTransaction())) {
            case "transfer":
            case "createToken":
            case "redeemToken":
                return new Transaction({
                    type: TransactionType.Transfer,
                    value: Transfer.tryFromTxn(txn),
                });
            case "initiateTransfer":
                return new Transaction({
                    type: TransactionType.InitiateTransfer,
                    value: Transfer.tryFromTxn(txn),
                });
            case "commitTransfer":
                return new Transaction({
                    type: TransactionType.CommitTransfer,
                    value: Transfer.tryFromTxn(txn),
                });
            case "createLedgerAccount":
            case "setInstrument":
            case "setBalanceLimit":
            case "setFreezeState":
                return new Transaction({
                    type: TransactionType.AccountUpdate,
                    value: AccountUpdate.tryFrom(txn),
                });
            case "invokeAction":
                return new Transaction({
                    type: TransactionType.Action,
                    value: Action.tryFromFinalizedTransaction(txn),
                });
            case "documentOperations":
                return new Transaction({
                    type: TransactionType.DocumentOperations,
                    value: null,
                });
        }
    }
}

interface IContextFilter extends ITxnFilter {
    min: TxId;
    max: TxId;
    limit: number;
    contextId: Uint8Array;
}

export class ContextFilter extends TxnFilter {
    private _contextId: Uint8Array;

    public constructor(properties: IContextFilter) {
        super(properties);
        this._contextId = properties.contextId;
    }

    public toListTransactionsRequest(): m10.sdk.ListTransactionsRequest {
        return new m10.sdk.ListTransactionsRequest({
            contextId: this._contextId,
            minTxId: this.getMinTxId(),
            maxTxId: this.getMaxTxId(),
            limit: this.getLimit(),
        });
    }

    public static byContextId(contextId: Uint8Array): ContextFilter {
        return new ContextFilter({
            min: 0,
            max: Number.MAX_SAFE_INTEGER,
            limit: TxnFilter.DEFAULT_TXN_LIMIT,
            contextId: contextId,
        });
    }
}


interface IGroupingFilter extends ITxnFilter {
    min: TxId;
    max: TxId;
    limit: number;
    accountId: AccountId;
}

export class GroupingFilter extends TxnFilter {
    private _accountId: AccountId;

    public constructor(properties: IGroupingFilter) {
        super(properties);
        this._accountId = properties.accountId;
    }

    public toGroupTransactionsRequest(): m10.sdk.GroupTransactionsRequest {
        return new m10.sdk.GroupTransactionsRequest({
            accountId: this._accountId.toUint8Array(),
            minTxId: this.getMinTxId(),
            maxTxId: this.getMaxTxId(),
            limitGroups: this.getLimit(),
        });
    }

    public static byAccountId(accountId: AccountId): GroupingFilter {
        return new GroupingFilter({
            min: 0,
            max: Number.MAX_SAFE_INTEGER,
            limit: TxnFilter.DEFAULT_TXN_LIMIT,
            accountId: accountId,
        });
    }
}
